common/sampling.cpp

```diff
@@ -0,0 +1,166 @@
+#include "sampling.h"
+
+llama_sampling_context::~llama_sampling_context() {
+    for (auto & it : sequence_contexts) {
+        if (it.second.grammar != NULL) {
+            llama_grammar_free(it.second.grammar);
+            it.second.grammar = NULL;
+        }
+    }
+}
+
+llama_sampling_context llama_sampling_context_init(
+        const struct gpt_params & params,
+                  llama_grammar * grammar) {
+  llama_sampling_context result;
+
+  result.params = params.sampling_params;
+  result.grammar = grammar;
+  return result;
+}
+
+// Note: Creates the context if it doesn't exist, so this always return something.
+llama_sampler_sequence_context & llama_sampling_get_sequence_context(
+              llama_sampling_context & sampling_ctx,
+        const llama_seq_id             seq) {
+    const auto it = sampling_ctx.sequence_contexts.find(seq);
+    if (it != sampling_ctx.sequence_contexts.end()) {
+        return it->second;
+    }
+    llama_sampler_sequence_context new_ctx = {
+        2.0f * sampling_ctx.params.mirostat_tau,
+        sampling_ctx.grammar != NULL ? llama_grammar_copy(sampling_ctx.grammar) : NULL,
+    };
+    return sampling_ctx.sequence_contexts.insert({seq, new_ctx}).first->second;
+}
+
+bool llama_sampling_context_reset(
+              llama_sampling_context & sampling_ctx,
+        const llama_seq_id             seq) {
+    const auto it = sampling_ctx.sequence_contexts.find(seq);
+    if (it == sampling_ctx.sequence_contexts.end()) return false;
+    if (it->second.grammar != NULL) {
+        llama_grammar_free(it->second.grammar);
+        it->second.grammar = NULL;
+    }
+    sampling_ctx.sequence_contexts.erase(it);
+    return true;
+}
+
+llama_token llama_sampling_sample(
+                  struct llama_context * ctx,
+                  struct llama_context * ctx_guidance,
+                  struct llama_sampling_context & sampling_ctx,
```

Comment:
```suggestion
                  struct llama_sampling_context & ctx_sampling,
```

There are a few other places that need similar change for consistency sake

Reply:
By "a few" you mean 20 or so? :) Hopefully I caught them all. Everything seems to compile/work still.